appscale/tools/agents/azure_agent.py

```diff
@@ -0,0 +1,694 @@
+    which is needed to access any resources.
+    Args:
+      parameters: A dict, containing all the parameters necessary to authenticate
+      this user with Azure. We assume that the user has already authorized this
+      account by creating a Service Principal with the appropriate (Contributor)
+      role.
+    Returns:
+      A ServicePrincipalCredentials instance, that can be used to access or
+      create any resources.
+    """
+    app_id = parameters[self.PARAM_APP_ID]
+    app_secret_key = parameters[self.PARAM_APP_SECRET]
+    tenant_id = parameters[self.PARAM_TENANT_ID]
+
+    # Get an Authentication token using ADAL.
+    context = adal.AuthenticationContext(self.AZURE_AUTH_ENDPOINT + tenant_id)
+    token_response = context.acquire_token_with_client_credentials(
+      self.AZURE_RESOURCE_URL, app_id, app_secret_key)
+    token_response.get('accessToken')
+
+    # To access Azure resources for an application, we need a Service Principal
+    # with the accurate role assignment. It can be created using the Azure CLI.
+    credentials = ServicePrincipalCredentials(client_id=app_id,
+                                              secret=app_secret_key,
+                                              tenant=tenant_id)
+    return credentials
+
+
+  def create_network_interface(self, network_client, region, group_name,
+    interface_name, network_name, subnet_name, ip_name):
+    """ A helper function that creates the network resources, such as virtual
+    network, public ip and network interface.
+    Args:
+      network_client: A NetworkManagementClient instance
+      region: The location specified in the AppScalefile
+      group_name: The Azure resource group name under which the VM is created.
+      interface_name: The name to use for the Network Interface resource.
+      network_name: The name to use for the Virtual Network resource.
+      subnet_name: The name to use for the Subnet resource.
+      ip_name: The name to use for the Public IP Address resource.
+    """
+    AppScaleLogger.log("Creating/Updating the Virtual Network '{}'".
+      format(network_name))
+    address_space = AddressSpace(address_prefixes=['10.1.0.0/16'])
+    subnet1 = Subnet(name=subnet_name, address_prefix='10.1.0.0/24')
+    network_client.virtual_networks.create_or_update(group_name, network_name,
+      VirtualNetwork(location=region, address_space=address_space,
+                     subnets=[subnet1]))
+
+    time.sleep(self.SLEEP_TIME)
+    subnet = network_client.subnets.get(group_name, network_name, subnet_name)
+
+    AppScaleLogger.log("Creating/Updating the Public IP Address '{}'".
+      format(ip_name))
+    ip_address = PublicIPAddress(
+      location=region, public_ip_allocation_method=IPAllocationMethod.dynamic,
+      idle_timeout_in_minutes=4)
+    network_client.public_ip_addresses.create_or_update(group_name, ip_name, ip_address)
+
+    time.sleep(self.SLEEP_TIME)
```

Comment:
The sleep is after creating the resources and before calling the 'get' method on them. The Network interface is dependent on the creation of the Public IP Address and that in turn is dependent on the Virtual Network. Without the sleep, it usually runs into resource not available error. But I did find that there was a polling function available, so used that to wait until done.